client/optimizing.go

```diff
@@ -0,0 +1,309 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/drand/drand/chain"
+	"github.com/drand/drand/log"
+)
+
+const (
+	defaultRequestTimeout     = time.Second * 5
+	defaultSpeedTestInterval  = time.Minute * 5
+	defaultRequestConcurrency = 2
+)
+
+// NewOptimizingClient creates a drand client that measures the speed of clients
+// and uses the fastest one.
+func NewOptimizingClient(
+	clients []Client,
+	chainInfo *chain.Info,
+	requestTimeout time.Duration,
+	requestConcurrency int,
+	speedTestInterval time.Duration,
+) (Client, error) {
+	if len(clients) == 0 {
+		return nil, errors.New("missing clients")
+	}
+	if chainInfo == nil {
+		return nil, errors.New("missing chain info")
+	}
+	stats := make([]*requestStat, len(clients))
+	now := time.Now()
+	for i, c := range clients {
+		stats[i] = &requestStat{client: c, rtt: 0, startTime: now}
+	}
+	done := make(chan struct{})
+	if requestTimeout <= 0 {
+		requestTimeout = defaultRequestTimeout
+	}
+	if requestConcurrency <= 0 {
+		requestConcurrency = defaultRequestConcurrency
+	}
+	if speedTestInterval <= 0 {
+		speedTestInterval = defaultSpeedTestInterval
+	}
+	oc := &optimizingClient{
+		clients:            clients,
+		stats:              stats,
+		requestTimeout:     requestTimeout,
+		requestConcurrency: requestConcurrency,
+		speedTestInterval:  speedTestInterval,
+		chainInfo:          chainInfo,
+		log:                log.DefaultLogger,
+		done:               done,
+	}
+	go oc.testSpeed()
+	return oc, nil
+}
+
+type optimizingClient struct {
+	sync.RWMutex
+	clients            []Client
+	stats              []*requestStat
+	requestTimeout     time.Duration
+	requestConcurrency int
+	speedTestInterval  time.Duration
+	chainInfo          *chain.Info
+	log                log.Logger
+	done               chan struct{}
+}
+
+type requestStat struct {
+	// client is the client used to make the request.
+	client Client
```

Comment:
maybe it makes more sense to have stats be in a map keyed by client?

Reply:
The stats are a slice because they’re ordered by speed. If they were a map we’d need to sort them when Get is called. Currently we sort them after we get a speed measurement as I figured in the happy path it would be faster to sort an already mostly sorted slice.

Also, on the happy path we’ll only get 1 new stat per call to Get, and the stat we want to update will be the first item in the stats slice. So there’s basically no looping.